parser_hevc.c

```diff
@@ -0,0 +1,387 @@
+// Tested with German DVB-T2 HD channels
+
+#include "parser_hevc.h"
+#include "bitstream.h"
+#include "config.h"
+
+#include <stdlib.h>
+#include <assert.h>
+
+
+cParserHEVC::cParserHEVC(int pID, cTSStream *stream, sPtsWrap *ptsWrap, bool observePtsWraps)
+ : cParser(pID, stream, ptsWrap, observePtsWraps)
+{
+  m_Height            = 0;
+  m_Width             = 0;
+  m_FpsScale          = 0;
+  m_PixelAspect.den   = 1;
+  m_PixelAspect.num   = 0;
+  memset(&m_streamData, 0, sizeof(m_streamData));
+  m_PesBufferInitialSize      = 240000;
+
+  m_IsVideo = true;
+  Reset();
+}
+
+cParserHEVC::~cParserHEVC()
+{
+}
+
+void cParserHEVC::Parse(sStreamPacket *pkt, sStreamPacket *pkt_side_data)
+{
+  if (m_PesBufferPtr < 10) // 2*startcode + header + trail bits
+    return;
+
+  int p = m_PesParserPtr;
+  uint32_t startcode = m_StartCode;
+  bool frameComplete = false;
+
+  while ((m_PesBufferPtr - p) > 3)
+  {
+    if ((startcode & 0x00ffffff) == 0x00000001)
+    {
+      if (Parse_HEVC(p, frameComplete) < 0)
+        break;
+    }
+    startcode = startcode << 8 | m_PesBuffer[p++];
+  }
+  m_PesParserPtr = p;
+  m_StartCode = startcode;
+
+  if (frameComplete)
+  {
+    if (!m_NeedSPS && m_FrameValid)
+    {
+      double PAR = (double)m_PixelAspect.num/(double)m_PixelAspect.den;
+      double DAR = (PAR * m_Width) / m_Height;
+      DEBUGLOG("HEVC SPS: PAR %i:%i", m_PixelAspect.num, m_PixelAspect.den);
+      DEBUGLOG("HEVC SPS: DAR %.2f", DAR);
+
+      int duration;
+      if (m_curDTS != DVD_NOPTS_VALUE && m_prevDTS != DVD_NOPTS_VALUE && m_curDTS > m_prevDTS)
+        duration = m_curDTS - m_prevDTS;
+      else
+        duration = m_Stream->Rescale(20000, 90000, DVD_TIME_BASE);
+
+      if (m_FpsScale == 0)
+        m_FpsScale = m_Stream->Rescale(duration, DVD_TIME_BASE, 90000);
+
+      bool streamChange = m_Stream->SetVideoInformation(m_FpsScale, DVD_TIME_BASE, m_Height, m_Width, DAR);
+
+      pkt->id       = m_pID;
+      pkt->size     = m_PesNextFramePtr;
+      pkt->data     = m_PesBuffer;
+      pkt->dts      = m_DTS;
+      pkt->pts      = m_PTS;
+      pkt->duration = duration;
+      pkt->streamChange = streamChange;
+
+    }
+    m_StartCode = 0xffffffff;
+    m_PesParserPtr = 0;
+    m_FoundFrame = false;
+    m_FrameValid = true;
+  }
+}
+
+void cParserHEVC::Reset()
+{
+  cParser::Reset();
+  m_StartCode = 0xffffffff;
+  m_NeedSPS = true;
+  m_NeedPPS = true;
+  memset(&m_streamData, 0, sizeof(m_streamData));
+  m_rbspBufferSize = 0;
+  if (m_rbspBuffer)
+  {
+    free(m_rbspBuffer);
+    m_rbspBuffer = NULL;
+  }
+}
+
+int cParserHEVC::Parse_HEVC(int buf_ptr, bool &complete)
+{
+  unsigned int len = m_PesBufferPtr - buf_ptr;
+  uint8_t *buf = m_PesBuffer + buf_ptr;
+  unsigned int p=0, NumBytesInNalUnit=0;
+
+  // find length of nal (including nal_unit_header)
+  uint32_t startcode = 0xffffffff;
+  while ((p+3) < len)
+  {
+    if ((startcode & 0x00ffffff) == 0x00000001)
```

Comment:
Sounds good. At least, we don't need to buffer the big VCL NALs anymore, only the ones that are actually parsed. If  I understand the spec right, the NAL header can never have the EP3Bytes, so:

1. Search next startcode
2. Read in NAL header and check unit code:
if non-parsed code (VCLs) then goto 1
3. Call StripEP3Bytes (which is actually already in the code, but for all NALs right now)
4. Hand over to cBitstream 
5. Parse NAL unit
This would still require a buffer, but quite small.

On the other hand, if cBitstream could take care of the EP3Bytes, then no buffer at all would be needed.
What I don't like is that cBitstream must look 24 bits ahead in order detect the extra 0x03 byte, which means that the last 16 bits in the stream have to handled separately.

So what should I implement: small buffer or enhanced cBitstream?